Sources/Plasma/PubUtilLib/plAudio/plVoiceCodec.cpp

```diff
@@ -0,0 +1,518 @@
+    bool                        fVBR;                   // toggle variable bit rate
+    int                         fAverageBitrate;        // n-bits per second
+    uint8_t                     fComplexity;            // 1-10 sets cpu resources allowed for encoder
+    bool                        fENH;                   // perceptual enhancement
+    bool                        fInitialized;
+};
+
+plSpeex::plSpeex() :
+    fBits(new SpeexBits),
+    fEncoderState(),
+    fDecoderState(),
+    fSampleRate(kSpeexSampleRate),
+    fFrameSize(-1),
+    fQuality(7),
+    fVBR(true),
+    fAverageBitrate(8000), // 8kb bitrate
+    fComplexity(3),
+    fENH(false),
+    fInitialized(false)
+{
+    Init();
+}
+
+plSpeex::~plSpeex()
+{
+    Shutdown();
+}
+
+bool plSpeex::Init()
+{
+    // setup speex
+    speex_bits_init(fBits.get());
+    fBitsInit = true;
+
+    fEncoderState = speex_encoder_init(speex_lib_get_mode(SPEEX_MODEID_NB));
+    fDecoderState = speex_decoder_init(speex_lib_get_mode(SPEEX_MODEID_NB));
+
+    speex_encoder_ctl(fEncoderState, SPEEX_GET_FRAME_SIZE, &fFrameSize);            // get frame size
+    speex_encoder_ctl(fEncoderState, SPEEX_SET_COMPLEXITY, &fComplexity);           // 3
+    speex_encoder_ctl(fEncoderState, SPEEX_SET_SAMPLING_RATE, &fSampleRate);        // 8 khz
+    speex_encoder_ctl(fEncoderState, SPEEX_SET_VBR_QUALITY, &fQuality);             // 7
+    speex_encoder_ctl(fEncoderState, SPEEX_SET_VBR, &fVBR);                         // use variable bit rate
+    speex_encoder_ctl(fEncoderState, SPEEX_SET_ABR, &fAverageBitrate);              // default to 8kb
+
+    speex_decoder_ctl(fDecoderState, SPEEX_SET_ENH, &fENH);                         // perceptual enhancement
+
+    fInitialized = true;
+
+    return true;
+}
+
+bool plSpeex::Shutdown()
+{
+    if (fDecoderState) {
+        speex_decoder_destroy(fDecoderState);
+        fDecoderState = nullptr;
+    }
+
+    if (fEncoderState) {
+        speex_encoder_destroy(fEncoderState);
+        fEncoderState = nullptr;
+    }
+
+    if (fBitsInit) {
+        speex_bits_destroy(fBits.get());
+        fBitsInit = false;
+    }
+    fInitialized = false;
+
+    return true;
+}
+
+bool plSpeex::Encode(const short* data, int numFrames, int& packedLength, void* out, int outsz)
+{
+    packedLength = 0;
+
+    const short* pData = data;                               // pointer to input data
+    std::unique_ptr<float> input{ new float[fFrameSize] };   // input to speex - used as am intermediate array since speex requires float data
+    uint8_t frameLength;                                     // number of bytes speex compressed frame to
+    std::unique_ptr<char> frameData{ new char[fFrameSize] }; // holds one frame of encoded data
+
+    // encode data
+    hsRAMStream stream;
+    for (int i = 0; i < numFrames; i++) {
+        // convert input data to floats
+        for (int j = 0; j < fFrameSize; j++)
+            input.get()[j] = pData[j];
```

Comment:
When using `unique_ptr<T[]>`, an `operator[]` is provided so you don't need to use `.get()`

Reply:
Kewl